### PR TITLE
civix#175 - Switch core extensions to mixins (ang-php, menu-xml, theme-php, case-xml, mgd-php)

### PR DESCRIPTION
**Pull request by totten**
Overview
---------

Building on #22198, this converts a bunch of boilerplate functions to use mixins. This removes ~3300 SLOC and (going forward) will allow other extensions to reduce SLOC.

Before
-------

Core extensions include boilerplate implementations of:

* `hook_angularModules` (`ang/*.ang.php`)
* `hook_caseTypes` (`xml/case/*.xml`)
* `hook_xmlMenu` (`xml/Menu/*.xml`)
* `hook_managed` (`*.mgd.php`)
* `hook_themes` (`*.theme.php`)

After
-------

The mixins `ang-php`, `case-xml`, `menu-xml`, `mgd-php`, and `theme-php` provide equivalent scanning behavior.

Some mixins (`case-xml`, `menu-xml`, `mgd-php`) have E2E test coverage (via `mixin/*/example` and `tools/mixin/bin/test-all`)

The boilerplate is removed.

Comments
------------

`hook_managed` is WIP - it regressed and is not currently passing in `tools/mixin/bin/test-all`.

### CODE REVIEW

**Comment by totten** [comment]
jenkins, test this please

**Comment by totten** [comment]
jenkins, test this please

**Comment by colemanw** [comment]
Retest this please

**Comment by totten** [comment]
@colemanw I've rebased and done some investigation on `hook_managed`. The problem that I'm having can be seen with these steps:

```bash
$ ./tools/mixin/bin/mixer create -f ext/myshimmy mgd-php@1
$ cd ext/myshimmy
$ phpunit8
```

(The `mixer create` command makes an extension with https://github.com/totten/civicrm-core/blob/master-mix-dec-all/mixin/mgd-php%401/example/CRM/ShimmyGroup.mgd.php.)

Or to reproduce manually:

```bash
## Create and install extension. It sets up the OptionGroup.
$ ./tools/mixin/bin/mixer create -f ext/myshimmy mgd-php@1
Create ext/myshimmy for mgd-php@1
$ cv en shimmy
Enabling extension "shimmy"
$ cv api4 OptionGroup.get +w name=shimmy_group
...Hooray, group exists...

## Remove the extension. It leaves behind a `civicrm_managed` record.
$ cv dis shimmy && cv ext:uninstall shimmy
Disabling extension "shimmy"
Uninstalling extension "shimmy"
$ echo 'select * from civicrm_managed' |  cv sql 
...
14	shimmy	the_managed_group	OptionGroup	108	NULL	NULL
...

## Re-install the extension. It fails to recreate the `OptionGroup`.
$ cv en shimmy
Enabling extension "shimmy"
$ cv api4 OptionGroup.get +w name=shimmy_group
...Oh no, no data!...
```

I guess I don't understand why it leaves behind a `civicrm_managed` record in this case? It seems like this is locking-in the deletion and preventint new `OptionGroup`s from being created?

**Comment by colemanw** [comment]
> I guess I don't understand why it leaves behind a `civicrm_managed` record in this case? It seems like this is locking-in the deletion and preventing new `OptionGroups` from being created?

@totten that's supposed to be happening here: https://github.com/civicrm/civicrm-core/blob/69e13f9b242901e272e8bf7e05a4f6dc9c57d01c/CRM/Core/BAO/Managed.php#L27-L34

**Comment by totten** [comment]
@colemanw  OK, so that function isn't called. Tracing `ManagedEntities::removeStaleEntity()`, it does call `civicrm_api4('OptionGroup', 'delete', ...)` - which eventually goes down to `CRM_Core_BAO_OptionGroup::del()`. However, it does not emit `hook_civicrm_post(delete, OptionGroup...)`.

I guess the situation is now like this: "If you enable `ManagedEntity` trait on an APIv4 entity, then the entity __must__ emit `hook_post`." (The `hook_post` behavior becomes mandatory rather than conventional.)

**Comment by totten** [comment]
Just some notes... Spent a while debugging an issue with using `mgd-php` for `search_kit` -- specifically, `SearchAfformTest` seems to fail. I *think* the problem has to do with the timing/ways in which `dispatcher` gets reset during the test lifecycle. [notes/logs](https://gist.github.com/totten/6d5490ed8fafc648abbea554e0919db8) In comparing the two logs:

* The biggest difference arises at the end -- after constructing `CiviEventDispatcher` twice, it does one final reconciliation. The data-dump during the last reconciliation shows that the `$managedEntities->declarations` is a much shorter list.
* There is also a difference in the order of items. But I've tried some hackery to change the order and didn't get anywhere with that, so it seems more likely that the other difference is more important.

**Comment by totten** [comment]
Well, the good news is that the refactoring above appears to have addressed the test failure. 

I want to add an explanatory note (which may not be useful to anyone else right now), but it took a while to figure, so it's good to leave the record. The old failure involving `mgd-php` and `search_kit` was a test-infra edge-case; in between test-cases, [`CiviTestListener` makes several efforts to reset global PHP state](https://github.com/civicrm/civicrm-core/blob/5.44.0/Civi/Test/CiviTestListenerPHPUnit7.php#L110-L114). It's hard to figure *why*, but the efforts are a bit redundant. It produce(d) a sequence of steps like this:

1. Destroy/recreate the `dispatcher`
2. Reinitialize mixins (ie register some listeners on `dispatcher`). Leave memos in `Civi::$statics` about which mixins were loaded.
3. Destroy/recreate the `dispatcher`
4. Reinitialize mixins - or try to.  The memos in `Civi::$statics` were *not* reset a second time, so it believed that mixins were already initialized - and it (incorrectly) chose to skip adding listeners.

In short, the lifecycles of `CiviEventDispatcher` and `MixinLoader` didn't quite match-up - so these semi-redundant resets impacted them differently.

The solution above was to rework the lifecycle of `MixinLoader` to make it *more boring* (more like other boot-services). So now it caches data in `CRM_Extension_System::singleton()->cache` (instead of `templates_c`) and it stores thread-local state in `$this` (instead of `Civi::$statics`).  The upshot - if you subject the system to multiple/redundant resets, it will behave more typically.

**Comment by totten** [comment]
Rebased/retesting again now that other/unrelated redflags should be cleared.

**Comment by colemanw** [comment]
Our old nemesis is back... "`Fatal: At least one command failed abnormally [ mixin]`"
Retest this please

**Comment by totten** [comment]
Ah, well, if we search the console log, it actually has a corresponding error when checking `case-xml@1`. It failed on both `testLifecycleWithSubprocesses` and `testLifecycleWithLocalFunctions`, but the latter backtrace looks clearer:

```
Create /home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/default/files/civicrm/ext/example-mixin for case-xml@1
Test case-xml@1
PHPUnit 8.5.15 by Sebastian Bergmann and contributors.

...

2) E2E_Shimmy_LifecycleTest::testLifecycleWithLocalFunctions
TypeError: array_merge(): Argument #1 must be of type array, null given

/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php:223
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php:149
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php:165
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/all/modules/civicrm/Civi/Api4/Generic/DAOGetAction.php:111
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/all/modules/civicrm/Civi/Api4/Generic/DAOGetAction.php:99
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/all/modules/civicrm/Civi/Api4/Provider/ActionObjectProvider.php:68
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/all/modules/civicrm/Civi/API/Kernel.php:149
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/all/modules/civicrm/Civi/Api4/Generic/AbstractAction.php:234
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/all/modules/civicrm/api/api.php:85
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:105
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/default/files/civicrm/ext/example-mixin/tests/mixin/CaseTypeTest.php:20
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:92
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:69
/home/jenkins/bknix-dfl/build/core-22199-3hy2j/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:50
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671

ERRORS!
Tests: 2, Assertions: 4, Errors: 2.
```

Need to figure out what's different (ie why it doesn't fail on my local...  and/or why it didn't fail before the rebase).

Will be interesting to see if the test on the later commit has that problem.

**Comment by seamuslee001** [comment]
@totten php versions?

**Comment by totten** [comment]
> (@seamuslee001) php versions?

It passes on my local with php 7.4.25 or 8.0.12. If I ssh into the failed build on `test-1` (running php 8.0.5), the failure does seem reproducible.

Checked a couple other hypotheses (which don't seem to explain either):

* On `test-1`, the problem arises with or without the recent CaseType BAO patch (11a1236cf36ca49443af1d644bdb42fd58ae180c).
* On `test-1`, the problem arises with or without the mixin refactoring (ie earlier 3774a0f and later 27b7127 are equally affected).

Still waiting to see what `test-3` says on its run.

**Comment by totten** [comment]
OK, the test discrepancy depends on CiviCase activation:

* On my regular dev-build, CiviCase is enabled (and this is baked into my local DB snapshot).
* On clean builds (*as in CI*), CiviCase is disabled - and one gets the error.

    ```
    civibuild create tmp --type drupal-clean --patch https://github.com/civicrm/civicrm-core/pull/22199
    ...
    cv api4 CaseType.get
    ...
    array_merge(): Argument #1 must be of type array, null given
    ```

Stepping through, it appears that `CaseType.get` starts executing but fails because there is no metadata for the `CaseType` entity. [This conditional](https://github.com/civicrm/civicrm-core/blob/f004c826f50cac33dd54f40a3d4e5f7fc2d869fe/Civi/Api4/Action/Entity/Get.php#L67) omits `CaseType`. So when it lookups metadata, it returns a `NULL` value -- and we ultimately get the error on [`array_merge(NULL...)`](https://github.com/civicrm/civicrm-core/blob/63f3df59810097f69b0c9503cb725d95c43d43f1/Civi/Api4/Query/Api4SelectQuery.php#L222-L223).

The quick fix is to just enable CiviCase during the E2E test. Pushed up a patch to do that.

Tangentially... split off an issue: https://lab.civicrm.org/dev/core/-/issues/2984

**Comment by totten** [comment]
Woot green @seamuslee001 @colemanw

**Comment by colemanw** [comment]
I've reviewed the code and it all looks sensible. Most of the changes are removing boilerplate and adding tests, which is great. Let's merge this into the RC so it can get more eyes on it over the next 4 weeks.
